File: golang/61.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func rotateRight(head *ListNode, k int) *ListNode {
    pCur := head
    n := 0
    for {
        if pCur == nil {
            break
        }
        
        n += 1
        pCur = pCur.Next
    }
    
    if n == 0 {
        return head
    }
    
    k %= n
    k = n - k
    
    if k == n {
        return head
    }
    
    pCur = head
    var pNext *ListNode = nil

    fmt.Println(k,n)
    for {
        if k == 1 {
            pNext = pCur.Next
            pCur.Next = nil
            break
        }
        pCur = pCur.Next
        k--
    }
    
    pCur = pNext
    
    for {
        if pCur.Next == nil {
            break
        }
        pCur = pCur.Next
    }
    
    pCur.Next = head
    
    return pNext
}

File: cpp/61.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if(head==NULL)return head;
        int cnt=0;
        ListNode*pCur=head;
        while(pCur)
        {
            cnt++;
            pCur=pCur->next;
        }
        k%=cnt;
        while(k--)
        {
            ListNode*pCur=head;
            int curVal=0,preVal=0;
            while(pCur)
            {
                curVal=pCur->val;
                pCur->val=preVal;
                preVal=curVal;
                pCur=pCur->next;
            }
            head->val=preVal;    
        }
        return head;
    }
};
